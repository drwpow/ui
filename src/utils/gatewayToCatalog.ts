import { Gateway } from '../types/gateway';
import { Catalog } from '../types/catalog';
import { globalRegion } from '../data/region';

export const convertPlan = (plan: Gateway.ResolvedPlan, productId: string, providerId: string): Catalog.Plan => ({
  id: plan.id || '',
  body: {
    ...plan,
    product_id: productId,
    provider_id: providerId,
    regions: [globalRegion.id], // TODO: Until we get graphQL, we're going to lose regions
  },
  version: 1,
  type: 'plan',
});

export const convertProduct = (product: Gateway.ResolvedProduct): Catalog.Product => ({
  id: product.id || '',
  version: 1,
  type: 'product',
  body: {
    ...product,
  },
});
