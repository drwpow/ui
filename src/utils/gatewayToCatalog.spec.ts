import { convertPlan, convertProduct } from './gatewayToCatalog';
import { ResolvedPlan, ResolvedProduct } from '../spec/mock/gateway';

describe('The convertPlan method', () => {
  it('will transform a gateway plan into a valid catalog plan', () => {
    const productId = ResolvedProduct.id || '';
    const providerId = ResolvedProduct.provider ? ResolvedProduct.provider.id : '';
    expect(convertPlan(ResolvedPlan, productId, providerId || '')).toMatchSnapshot();
  });
});

describe('The convertProduct method', () => {
  it('will transform a gateway product into a valid catalog product', () => {
    expect(convertProduct(ResolvedProduct)).toMatchSnapshot();
  });
});
