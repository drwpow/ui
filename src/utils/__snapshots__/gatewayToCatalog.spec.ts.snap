// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The convertPlan method will transform a gateway plan into a valid catalog plan 1`] = `
Object {
  "body": Object {
    "cost": 500,
    "defaultCost": 500,
    "expanded_features": Array [
      Object {
        "label": "region",
        "name": "Region",
        "type": "string",
        "value": Object {
          "label": "us-east-1",
          "name": "US East 1",
        },
        "values": Array [
          Object {
            "label": "us-east-1",
            "name": "US East",
          },
          Object {
            "label": "us-west-1",
            "name": "US West",
          },
          Object {
            "label": "ap-south-1",
            "name": "Asia Pacific (Mumbai)",
          },
          Object {
            "label": "ca-central-1",
            "name": "Canada (Central)",
          },
          Object {
            "label": "us-west-1",
            "name": "EU (Ireland)",
          },
        ],
      },
      Object {
        "label": "storage",
        "name": "Storage",
        "type": "number",
        "value": Object {
          "label": "storage",
          "name": "5 GB",
          "numeric_details": Object {
            "cost_ranges": Array [
              Object {
                "limit": 5,
              },
              Object {
                "cost_multiple": 200000000,
                "limit": 16000,
              },
            ],
            "increment": 1,
            "max": 16000,
            "min": 5,
            "suffix": "GB",
          },
          "price": Object {
            "description": "Free",
            "formula": "(* storage#cost backups#number)",
          },
        },
        "value_string": "Storage",
        "values": Array [
          Object {
            "label": "storage",
            "name": "5 GB",
            "numeric_details": Object {
              "cost_ranges": Array [
                Object {
                  "limit": 5,
                },
                Object {
                  "cost_multiple": 200000000,
                  "limit": 16000,
                },
              ],
              "increment": 1,
              "max": 16000,
              "min": 5,
              "suffix": "GB",
            },
            "price": Object {
              "description": "Free",
              "formula": "(* storage#cost backups#number)",
            },
          },
        ],
      },
      Object {
        "label": "redundancy",
        "name": "High Availability",
        "type": "boolean",
        "value": Object {
          "label": "false",
          "name": "false",
        },
        "value_string": "No",
        "values": Array [
          Object {
            "label": "true",
            "name": "Yes",
            "price": Object {
              "description": "Creates a clone of the DB to rescue from potential outages",
              "formula": "(* plan#partial_cost redundancy#multiply_factor)",
              "multiply_factor": 0.8,
            },
          },
          Object {
            "label": "false",
            "name": "No",
          },
        ],
      },
    ],
    "features": Array [
      Object {
        "feature": "region",
        "value": "us-east-1",
      },
      Object {
        "feature": "storage",
        "value": "storage",
      },
      Object {
        "feature": "redundancy",
        "value": "false",
      },
    ],
    "free": true,
    "id": "235abe2ba8b39e941u2h70ayw5m9j",
    "label": "kitefin",
    "name": "Kitefin",
    "product_id": "234w1jyaum5j0aqe3g3bmbqjgf20p",
    "provider_id": "2346mdxcuca9ez2n93f72nb2fpjgu",
    "regions": Array [
      "235n4f9pxf8eyraj3y159x89z6jer",
    ],
    "resizable_to": Array [
      "23556xtwybe9zv877e6g2c7ygnxpj",
      "2358u3xj4d8x5qud8x1atgafwchfy",
    ],
    "state": "available",
    "trial_days": 0,
  },
  "id": "235abe2ba8b39e941u2h70ayw5m9j",
  "type": "plan",
  "version": 1,
}
`;

exports[`The convertProduct method will transform a gateway product into a valid catalog product 1`] = `
Object {
  "body": Object {
    "billing": Object {
      "currency": "usd",
      "type": "monthly-prorated",
    },
    "documentation_url": "https://jawsdb.com/docs",
    "feature_types": Array [
      Object {
        "label": "region",
        "name": "Region",
        "type": "string",
        "value": Object {
          "label": "us-east-1",
          "name": "US East 1",
        },
        "values": Array [
          Object {
            "label": "us-east-1",
            "name": "US East",
          },
          Object {
            "label": "us-west-1",
            "name": "US West",
          },
          Object {
            "label": "ap-south-1",
            "name": "Asia Pacific (Mumbai)",
          },
          Object {
            "label": "ca-central-1",
            "name": "Canada (Central)",
          },
          Object {
            "label": "us-west-1",
            "name": "EU (Ireland)",
          },
        ],
      },
      Object {
        "customizable": true,
        "downgradable": true,
        "label": "instance_class",
        "name": "RAM",
        "type": "string",
        "upgradable": true,
        "value": Object {
          "label": "db.t2.micro",
          "name": "1 GB",
          "price": Object {},
        },
        "value_string": "1 GB",
        "values": Array [
          Object {
            "label": "db.t2.micro",
            "name": "1 GB",
            "price": Object {},
          },
          Object {
            "cost": 5400,
            "label": "db.t2.small",
            "name": "2 GB",
            "price": Object {
              "cost": 5400,
            },
          },
          Object {
            "cost": 13400,
            "label": "db.m3.medium",
            "name": "4 GB",
            "price": Object {
              "cost": 13400,
            },
          },
          Object {
            "cost": 31800,
            "label": "db.m4.large",
            "name": "8 GB",
            "price": Object {
              "cost": 31800,
            },
          },
          Object {
            "cost": 54400,
            "label": "db.r4.large",
            "name": "15 GB",
            "price": Object {
              "cost": 54400,
            },
          },
          Object {
            "cost": 89000,
            "label": "db.r4.xlarge",
            "name": "31 GB",
            "price": Object {
              "cost": 89000,
            },
          },
          Object {
            "cost": 121000,
            "label": "db.r4.2xlarge",
            "name": "61 GB",
            "price": Object {
              "cost": 121000,
            },
          },
          Object {
            "cost": 239000,
            "label": "db.r4.4xlarge",
            "name": "122 GB",
            "price": Object {
              "cost": 239000,
            },
          },
          Object {
            "cost": 428000,
            "label": "db.r4.8xlarge",
            "name": "244 GB",
            "price": Object {
              "cost": 428000,
            },
          },
          Object {
            "cost": 910000,
            "label": "db.r4.16xlarge",
            "name": "488 GB",
            "price": Object {
              "cost": 910000,
            },
          },
        ],
      },
      Object {
        "label": "storage",
        "name": "Storage",
        "type": "number",
        "value": Object {
          "label": "storage",
          "name": "5 GB",
          "numeric_details": Object {
            "cost_ranges": Array [
              Object {
                "limit": 5,
              },
              Object {
                "cost_multiple": 200000000,
                "limit": 16000,
              },
            ],
            "increment": 1,
            "max": 16000,
            "min": 5,
            "suffix": "GB",
          },
          "price": Object {
            "description": "Free",
            "formula": "(* storage#cost backups#number)",
          },
        },
        "value_string": "Storage",
        "values": Array [
          Object {
            "label": "storage",
            "name": "5 GB",
            "numeric_details": Object {
              "cost_ranges": Array [
                Object {
                  "limit": 5,
                },
                Object {
                  "cost_multiple": 200000000,
                  "limit": 16000,
                },
              ],
              "increment": 1,
              "max": 16000,
              "min": 5,
              "suffix": "GB",
            },
            "price": Object {
              "description": "Free",
              "formula": "(* storage#cost backups#number)",
            },
          },
        ],
      },
      Object {
        "label": "hourly-price",
        "measurable": true,
        "name": "Hourly Price",
        "type": "number",
        "value": Object {
          "label": "trial",
          "name": "Trial GPU",
          "numeric_details": Object {
            "cost_ranges": Array [
              Object {
                "limit": 10,
              },
            ],
            "increment": 1,
            "suffix": "hour",
          },
        },
        "value_string": "NVIDIA 1080ti 250 GB SSD",
        "values": Array [
          Object {
            "label": "trial",
            "name": "Trial GPU",
            "numeric_details": Object {
              "cost_ranges": Array [
                Object {
                  "limit": 10,
                },
              ],
              "increment": 1,
              "suffix": "hour",
            },
          },
          Object {
            "label": "too-expensive-4u",
            "name": "No GPU",
            "numeric_details": Object {
              "cost_ranges": Array [],
              "increment": 1,
              "suffix": "hour",
            },
          },
          Object {
            "label": "tiered-plan",
            "name": "NVIDIA Tiered PLan",
            "numeric_details": Object {
              "cost_ranges": Array [
                Object {
                  "limit": 10,
                },
                Object {
                  "cost_multiple": 900000000,
                  "limit": 100,
                },
                Object {
                  "cost_multiple": 250000000,
                  "limit": -1,
                },
                Object {
                  "cost_multiple": 300000000,
                  "limit": 400,
                },
                Object {
                  "cost_multiple": 500000000,
                  "limit": 300,
                },
                Object {
                  "cost_multiple": 600000000,
                  "limit": 200,
                },
              ],
              "increment": 1,
              "suffix": "hour",
            },
          },
        ],
      },
      Object {
        "customizable": true,
        "label": "storage",
        "name": "Storage",
        "type": "number",
        "upgradable": true,
        "value": Object {
          "label": "storage",
          "name": "5 GB",
          "numeric_details": Object {
            "cost_ranges": Array [
              Object {
                "limit": 5,
              },
              Object {
                "cost_multiple": 200000000,
                "limit": 16000,
              },
            ],
            "increment": 1,
            "max": 16000,
            "min": 5,
            "suffix": "GB",
          },
          "price": Object {
            "description": "Free",
            "formula": "(* storage#cost backups#number)",
          },
        },
        "value_string": "Storage",
        "values": Array [
          Object {
            "label": "storage",
            "name": "5 GB",
            "numeric_details": Object {
              "cost_ranges": Array [
                Object {
                  "limit": 5,
                },
                Object {
                  "cost_multiple": 200000000,
                  "limit": 16000,
                },
              ],
              "increment": 1,
              "max": 16000,
              "min": 5,
              "suffix": "GB",
            },
            "price": Object {
              "description": "Free",
              "formula": "(* storage#cost backups#number)",
            },
          },
        ],
      },
      Object {
        "label": "redundancy",
        "name": "High Availability",
        "type": "boolean",
        "value": Object {
          "label": "false",
          "name": "false",
        },
        "value_string": "No",
        "values": Array [
          Object {
            "label": "true",
            "name": "Yes",
            "price": Object {
              "description": "Creates a clone of the DB to rescue from potential outages",
              "formula": "(* plan#partial_cost redundancy#multiply_factor)",
              "multiply_factor": 0.8,
            },
          },
          Object {
            "label": "false",
            "name": "No",
          },
        ],
      },
      Object {
        "customizable": true,
        "downgradable": true,
        "label": "redundancy",
        "name": "High Availability",
        "type": "boolean",
        "upgradable": true,
        "value": Object {
          "label": "false",
          "name": "false",
        },
        "value_string": "No",
        "values": Array [
          Object {
            "label": "true",
            "name": "Yes",
            "price": Object {
              "description": "Creates a clone of the DB to rescue from potential outages",
              "formula": "(* plan#partial_cost redundancy#multiply_factor)",
              "multiply_factor": 0.8,
            },
          },
          Object {
            "label": "false",
            "name": "No",
          },
        ],
      },
    ],
    "id": "234w1jyaum5j0aqe3g3bmbqjgf20p",
    "images": Array [
      "https://cdn.manifold.co/providers/jawsdb/screenshots/ss1.PNG",
      "https://cdn.manifold.co/providers/jawsdb/screenshots/ss2.PNG",
      "https://cdn.manifold.co/providers/jawsdb/screenshots/ss3.PNG",
      "https://cdn.manifold.co/providers/jawsdb/screenshots/ss4.PNG",
    ],
    "integration": Object {
      "base_url": "https://api.jawsdb.com/mysql",
      "features": Object {
        "access_code": false,
        "plan_change": true,
        "region": "user-specified",
        "sso": true,
      },
      "provisioning": "public",
      "sso_url": "https://api.jawsdb.com/mysql",
      "version": "v1",
    },
    "label": "jawsdb-mysql",
    "listing": Object {
      "listed": true,
      "marketing": Object {
        "beta": false,
        "featured": false,
        "new": false,
      },
      "public": true,
    },
    "logo_url": "https://cdn.manifold.co/providers/jawsdb/logos/80ca8b9113cf76fd.png",
    "name": "JawsDB MySQL",
    "provider": Object {
      "id": "2346mdxcuca9ez2n93f72nb2fpjgu",
      "label": "jawsdb",
      "name": "JawsDB",
    },
    "provider_id": "2346mdxcuca9ez2n93f72nb2fpjgu",
    "state": "available",
    "support_email": "support@jawsdb.com",
    "tagline": "Fast, reliable, no-bullshark MySQL as a Service",
    "tags": Array [
      "database",
    ],
    "terms": Object {
      "provided": false,
      "url": "",
    },
    "value_props": Array [
      Object {
        "body": "Get access to the same database trusted by sites such as Google, Facebook, Twitter, Youtube, and more.",
        "header": "Trust in Jaws",
      },
      Object {
        "body": "In addition to any backups you personally create, we automatically take nightly snapshots for single tenant databases. This means that our team can help you restore to a previous point in time.",
        "header": "Backups",
      },
      Object {
        "body": "Create, edit, and save reusable queries against your provisioned server. Databites are a custom, zero-hassle reporting solution for your database right out of the box!",
        "header": "Databites",
      },
      Object {
        "body": "Regardless of how big of a wave you're riding, JawsDB makes it easy to scale your database performance.",
        "header": "Scale",
      },
      Object {
        "body": "For single tenant plans, we provide you and only you with root credentials — so you have complete control of your server.",
        "header": "Stay in Control",
      },
      Object {
        "body": "The world doesn't need to be a scary place, choose a plan that supports failover and your data is replicated to servers in different regions of the world, reducing unexpected downtime.",
        "header": "Data Replication",
      },
    ],
  },
  "id": "234w1jyaum5j0aqe3g3bmbqjgf20p",
  "type": "product",
  "version": 1,
}
`;
